azext_iot/iothub/providers/device_identity.py

```diff
@@ -0,0 +1,426 @@
+        # device configs by id
+        device_config_dict: Dict[str, EdgeDevicesConfig] = {}
+        for device_config in config.devices:
+            device_config_dict[device_config.device_id] = device_config
+
+        # first pass to create devices in flat tree
+        config_devices_iterator = (
+            tqdm(
+                config.devices,
+                desc="Creating device structure and certificates",
+            )
+            if visualize
+            else config.devices
+        )
+        for device_config in config_devices_iterator:
+            device_id = device_config.device_id
+            # add to flat tree
+            try:
+                tree.create_node(device_id, device_id, parent=tree_root_node_id)
+            except DuplicatedNodeIdError:
+                raise InvalidArgumentValueError(
+                    f"Duplicate deviceId '{device_id}' detected"
+                )
+
+        # second pass to move nodes and check hierarchy
+        for device_config in config.devices:
+            device_id = device_config.device_id
+
+            # Move nodes to their correct parents, track device->parent in dict
+            device_parent = device_config.parent_id or tree_root_node_id
+            if device_parent != tree_root_node_id:
+                device_to_parent_dict[device_id] = device_parent
+            try:
+                tree.update_node(device_id, data=device_parent)
+                tree.move_node(device_id, device_parent)
+            except NodeIDAbsentError:
+                raise InvalidArgumentValueError(
+                    f"Error building device hierarchy, missing parent '{device_parent}'"
+                )
+            except LoopError:
+                raise InvalidArgumentValueError(
+                    "Error building device hierarchy, found a loop between "
+                    f"devices '{device_id}' and '{device_parent}'."
+                )
+
+        # Show the device tree
+        if visualize:
+            tree.show()
+
+        # Query existing devices
+        query_args = ["SELECT deviceId FROM devices"]
+        query_method = self.service_sdk.query.get_twins
+        existing_devices = _execute_query(query_args, query_method)
+        existing_device_ids = list(map(lambda x: x["deviceId"], existing_devices))
+
+        # Clear devices if necessary
+        if clean and len(existing_device_ids):
+            delete_iterator = (
+                tqdm(existing_device_ids, "Deleting existing device identities")
+                if visualize
+                else existing_device_ids
+            )
+            self.delete_device_identities(delete_iterator, confirm=True)
+        else:
+            # If not cleaning the hub, ensure no duplicate device ids
+            duplicates = list(
+                filter(lambda id: id in device_config_dict, existing_device_ids)
+            )
+            if any(duplicates):
+                raise InvalidArgumentValueError(
```

Comment:
not sure if there is a test for this yet

Reply:
Have a note to address this in a follow-up PR